src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+(*
+ * Copyright (c) 2022-2022 Tarides <contact@tarides.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+open Import
+include Chunked_suffix_intf
+
+module Make (Io : Io.S) (Errs : Io_errors.S with module Io = Io) = struct
+  module Io = Io
+  module Errs = Errs
+  module Ao = Append_only_file.Make (Io) (Errs)
+
+  type chunk = { idx : int; start_off : int63; ao : Ao.t }
+
+  let chunk_off_to_poff c off = Int63.Syntax.(off - c.start_off)
+
+  type create_error = Io.create_error
+
+  type open_error =
+    [ Io.open_error
+    | `Closed
+    | `Invalid_argument
+    | `Inconsistent_store
+    | `Read_out_of_bounds ]
+
+  (** A simple container for chunks. *)
+  module Inventory : sig
+    type t
+
+    val v : int -> (int -> chunk) -> t
+    val iter_skip_appendable : (chunk -> unit) -> t -> unit
+    val appendable : t -> chunk
+    val find : off:int63 -> t -> chunk
+
+    val open_ :
+      start_idx:int ->
+      chunk_num:int ->
+      open_chunk:
+        (chunk_idx:int -> is_appendable:bool -> (Ao.t, open_error) result) ->
+      (t, [> open_error ]) result
+  end = struct
+    type t = chunk Array.t
+
+    exception OpenInventoryError of open_error
+
+    let v = Array.init
+
+    let iter_skip_appendable f t =
+      Array.sub t 0 (Array.length t - 1) |> Array.iter f
+
+    let appendable t = Array.get t (Array.length t - 1)
+
+    let find ~off t =
+      let find c =
+        let end_poff = Ao.end_poff c.ao in
```

Comment:
again, I don't think this `Ao.end_poff` is want you should use here

If I understand correctly, this abstraction introduces a new layer of offsets. We now have:
- global offsets (dispatched between prefix and suffix);  
- suffix offsets (physical offsets in the old suffix)  -- the one you manipulate in most of this file, right?
- physical chunk offsets (the one returned by `Ao.end_poff` for instance).

Reply:
I think this one is fine because it's subtracting the chunk `start_off` in `chunk_off_to_poff` before comparing (?)

Reply:
Yes, just saw that, thanks!

Reply:
@icristescu your breakdown of the mapping of offsets is correct. A `Chunked_suffix` maps suffix offset to chunk physical offset (which each chunk storing its suffix offset to facilitate lookup).

Based on this discussion, I think the names could be improved:

- `chunk.start_off` could be `chunk.suffix_off`
- `chunk_off_to_poff` could be `suffix_off_to_chunk_poff` (verbose but clear (?))

WDYT?

Reply:
Made these name changes in [`c55a987` (#2115)](https://github.com/mirage/irmin/pull/2115/commits/c55a98774e1741a41021f6619e3581c2abf8ac66) and I think it improves the clarity!